pubtools/_pulp/tasks/garbage_collect.py

```diff
@@ -96,13 +90,45 @@ def clean_all_rpm_content(self):
             LOG.info("No repos found for cleaning.")
             return
         arc_repo = clean_repos[0]
-        deleted_arc = list(arc_repo.remove_content(criteria=criteria))
+
+        search_criteria = Criteria.and_(
+            Criteria.with_unit_type(RpmUnit, unit_fields=["unit_id"]),
+            Criteria.with_field(
+                "cdn_published",
+                Matcher.less_than(datetime.utcnow() - timedelta(days=arc_threshold)),
+            ),
+        )
         deleted_content = []
-        for task in deleted_arc:
-            if task.repo_id == "all-rpm-content":
-                for unit in task.units:
-                    LOG.info("Old all-rpm-content deleted: %s", unit.name)
-                    deleted_content.append(unit)
+
+        LOG.info("Collecting unit batches for deletion")
+        first_page = arc_repo.search_content(criteria=search_criteria)
+
+        def handle_pages(page):
+            batch_accumulator = [[unit.unit_id for unit in page.data]]
+            if page.next:
+                batch_accumulator.extend(f_flat_map(page.next, handle_pages).result())
```

Comment:
We'll need at least one test which generates more than one page of data so this will be covered.

Incidentally, I'm not sure there is much point of writing the code this way with flat_map because you immediately call 'result()` on the handler for the next page, making the whole thing blocking anyway, and making the call stack N deep if there are N pages.  So this is kind of paying the cost of having more complicated code dealing with futures, without getting the benefit as every future is immediately awaited and nothing is allowed to run concurrently.

This simpler approach might have worked (haven't tested it) and should allow some things to run concurrently:

```python
# Wait for the first page
page = arc_repo.search_content(criteria=search_criteria).result()

# While there is a page of data to be processed...
while page:
  # process it
  batch = page.data
  do_something_with(batch)

  # Wait for the next page (if there is one).
  # note: the client has been working on fetching the next page in the background
  # even before we get here.
  page = page.next.result() if page.next else None
```

Or, another alternative:

```python
batch = []
batch_size = 1000
for unit in arc_repo.search_content(criteria=search_criteria):
  batch.append(unit)
  if len(batch) > batch_size:
    process_batch(batch)
    batch = []
if batch:
  process_batch(batch)
```

Reply:
I'd like to process the search results into batches and then do the deletion afterwards. The search_content pagination seems to make requests with the limit/ skip values, so mixing  in deletion requests with that feels dangerous. I've simplified code to arrange the batches and added in a test.